### DISCUSSION TITLE
Tessie to TeslaMate Migration

### DISCUSSION THREAD

**Discussion by giterdone509** [teslamate-org/teslamate#4112] [Ideas]
Former Tessie user. I would love a feature to migrate my old Tessie data to TeslaMate. Tessie has a function to export data to a csv for import to TeslaFi.

**Comment by cwanja** [comment]
Could go Tessie > TeslaFi > TeslaMate since TeslaFi importing is supported.

**Reply by giterdone509** [reply]
I was incorrect with my statement. You can go TeslaFi to Tessie, but not the other way around.

**Comment by DrMichael** [comment]
If I could get a sample output, I would have a look...

**Reply by giterdone509** [reply]
Attached the drives and charges csv. Removed my location data. Let me know if you need anything further. 
[VIN-drives 2.csv](https://github.com/user-attachments/files/16462485/VIN-drives.2.csv)
[VIN-charges.csv](https://github.com/user-attachments/files/16462486/VIN-charges.csv)

**Comment by DrMichael** [comment]
I don't see any location data in them. So, tables `positions` and `addresses` would be empty, also `geofences`, `states` and `updates`.

And, not sure, how to deal with `charging_processes` and `charges`.

So, migration probably does not make sense.

**Reply by giterdone509** [reply]
There is location data. I removed the data in the columns for privacy. Do you have an email or something that I can send the file with location data to? 
![image](https://github.com/user-attachments/assets/a820ab3b-42c3-4786-a7ab-4dc18dcb3ad7)

**Comment by DrMichael** [comment]
Drop it here (link removed) and tell me, when done please.

**Comment by DrMichael** [comment]
So, I took a closer look at the drives data. One could fill in the columns of drives and positions partially (the blue fields come from Tessie):
![image](https://github.com/user-attachments/assets/88a9c8a8-0dcb-4b52-961f-b7f593e1092c)

Not sure, what one could get then in the dashboards though. Any thoughts?

(I am unsure about the 'location' in the Tessie data. What is it?)

**Reply by cwanja** [reply]
Feel like the juice is not worth the squeeze there. More negative downstream impacts by having partial data.

**Reply by giterdone509** [reply]
Uploaded everything to dropbox. Let me know if this helps fill in some of the gaps. Will trust your discretion if this would provide benefit to TeslaMate. From my POV, I have a bunch of data sitting in Tessie that I would love to migrate in to get a more comprehensive view of my all my data.

**Comment by DrMichael** [comment]
Well, obviously a lot of information, which teslamate collects, will be missing. E.g. teslamate records the positions during a drive. With this data, we will have only start and end of a drive...
I twiddled with the drives data. I used a fake car... The car should definitely be already in teslamate before importing.
Well, some results:
Drives:
![image](https://github.com/user-attachments/assets/9bdd2e33-d97a-4d11-a95f-7a88214f597a)
Drive Stats:
![image](https://github.com/user-attachments/assets/eba46b08-d590-4ee2-85ef-6fa97f426bfd)
Charges:
![image](https://github.com/user-attachments/assets/1a6e145f-1b6c-4937-839f-1e8dca7a67bf)
Charing Stats:
![image](https://github.com/user-attachments/assets/e0decd98-8ed0-4100-ae23-08a79002b525)
Statistics:
![image](https://github.com/user-attachments/assets/2a9bc347-c257-45f8-b856-fd18ccfa01f1)
Overview:
![image](https://github.com/user-attachments/assets/220fb8aa-64e8-4e10-9019-080dc2260f6e)

Not sure, whether it is worth ist...

**Reply by giterdone509** [reply]
The Tessie devs must be listening. They have changed the available export options. The files I sent you are no longer available. They now just have an export all function that exports all data. The new data includes files that report on the states. Seemingly this state data is what you're referring to (recording the position during the drive). Apologies for the work you have done with the old exports. Hopefully this data positions you better for a more comprehensive data migration. I have uploaded the new files to your dropbox as a zip. Really appreciate your work on this! 
![image](https://github.com/user-attachments/assets/83987212-473b-4cd5-9403-65b4ec06640c)

**Comment by DrMichael** [comment]
Hmmm, that data is not usable. No way to get the necessary informations for charges and drives out os that. (At least no obvious...)

Can't the Tessi guys provide a teslamate export? I attach the ER model, which should help with that.
![teslamate - ER diagram](https://github.com/user-attachments/assets/ffa92aa2-0d3f-44c0-a43f-9df219105823)

**Reply by edwinstol** [reply]
I would not expect support from Tessie on getting the data our of their systems and into TeslaMate. The other way around is supported by them (https://help.tessie.com/article/70-import-from-teslamate) since they would like people to start using Tessie and stop with TeslaMate, not the other way around.

**Comment by jlestel** [comment]
I understand that Tessie does not offer any export, it is up to us to work. And when I see what they export, I confirm that Teslamate is much more complete in terms of data!!

To import third-party data, I base myself on the format required for TeslaFi import. So I have prepared a CSVFiddle to show how we can transform the CSV files provided by Tessie with an SQL query that will generate a "TeslaFi" type file.

**Comment by jlestel** [comment]
I made a grinder to compile the Tessie files and create a CSV import file compatible with Teslamate. It is based on the TeslaFi format.

It is available here if you want to try: https://www.myteslamate.com/tessie-to-teslamate/

The data imports correctly, I've seen drives but I have not yet verified the completeness.

You can see the BigQuery query that creates the final file [here](https://github.com/MyTeslaMate/tessie-to-teslamate-import/blob/master/bigquery.sql)

**Reply by giterdone509** [reply]
Just gave it a try. Receiving this error 
![image](https://github.com/user-attachments/assets/07d257b9-111e-4f0a-b0e2-cab62ea39d2c)

**Reply by jlestel** [reply]
Can you send me a wetransfer link of your Tessie export in the chat?

**Reply by jlestel** [reply]
@giterdone509 Fixed

**Reply by lipov3cz3k** [reply]
I tried your convertor, but it ends with this error
```
Query error : { "error": { "code": 404, "message": "Not found: Table tessie-438008:LRW3E7EK2RC146216.charging_states was not found in location US", "errors": [ { "message": "Not found: Table tessie-438008:LRW3E7EK2RC146216.charging_states was not found in location US", "domain": "global", "reason": "notFound" } ], "status": "NOT_FOUND" } }
Fail to read date start / end
```
I am from EU, so maybe BigQuery is not available in my region?

**Reply by jlestel** [reply]
charging_states.csv was not found. You needed to create an archive ass described on website.

**Reply by ptiagosequeira** [reply]
I get the same error. charging_states.csv exists but fails to read the start date of charging
Thanks

`Query error : { "error": { "code": 404, "message": "Not found: Table tessie-438008:LRW3E7EK2MC326451.charging_states was not found in location US", "errors": [ { "message": "Not found: Table tessie-438008:LRW3E7EK2MC326451.charging_states was not found in location US", "domain": "global", "reason": "notFound" } ], "status": "NOT_FOUND" } }
Fail to read date start / end`

**Reply by luschmidt93** [reply]
I have exactly the same issue, tried everything with the charging_states.csv... the original one, an empty one, one with just a few entries, but remains the same. Any suggestions?

**Reply by jlestel** [reply]
Ask on the myteslamate.com chat with your VIN.

**Reply by lipov3cz3k** [reply]
I created my own google BigQuery and use linked sql query to convert my data and it worked fine.

**Reply by Yeraze** [reply]
Does this tool still work? i'm trying to use it with a Tessie export of about 3 years of data, and it's failing.. It gets to the first date export, and then I see the JS console die with `tessie:446 Error fetching progress: TypeError: Cannot read properties of null (reading 'split')
    at tessie:421:51`

**Reply by mapleroyal** [reply]
@Yeraze Did you ever get it working? It supposedly converts mine with no errors, but almost all the useful fields are blank strings. After importing into teslamate, every dashboard page just shows "Not Found".

@jlestel Does this tool still work? See my comment to Yeraze ^

**Comment by giterdone509** [comment]
Amazing! I will give it a try.

On Oct 8, 2024, at 10:50, jlestel ***@***.***> wrote:

﻿

I made a grinder to compile the Tessie files and create a CSV import file compatible with Teslamate. It is based on the TeslaFi format.

It is available here if you want to try: https://www.myteslamate.com/tessie-to-teslamate/

The data imports correctly, I've seen drives but I have not yet verified the completeness.

You can see the BigQuery query that creates the final file here<https://github.com/MyTeslaMate/tessie-to-teslamate-import/blob/master/bigquery.sql>

**Comment by davidstosik** [comment]
Hello @giterdone509. I've been excited to try your migration tool, since I have been tracking my Tesla from about 10 days after its delivery, last year, but would be interested in switching to TeslaMate.
Unfortunately, I'm hitting a snag as the migration tool is telling me "Invalid charging_states." Could it be that Tessie's export format changed? Or maybe it produces some data that your script is not expecting?

I noticed that "Charger States" is always null or `"1"`, even though I've charged in may different conditions (mobile charger, both 100V and 200V, Wall Connector, Supercharger...). I tried replacing null values with `"1"`, but the script still did not accept my data.
I also tried deleting timestamp duplicates, but unfortunately this didn't work either.

I can try sharing my `charging_states.csv` file with you if it helps, though it's 45MB.
I'd also be happy to help debugging, if you have any ideas in mind.

**Reply by jlestel** [reply]
Hi David, apparently Tesla changed the format on January 21st, I'll look into that.

**Reply by jlestel** [reply]
Hi @davidstosik it's fixed now!

**Comment by davidstosik** [comment]
@jlestel Thanks a lot, it worked! ❤️ 

<img width="296" alt="image" src="https://github.com/user-attachments/assets/016c5f4f-f12b-4b4e-a417-ab1d5a149c1d" />

**Comment by sbjensen** [comment]
Is the import broken?
I'm running version 2.0.0 of Teslamate and Postgres version 17.

Tried creating the Teslamate import csv file using this link https://app.myteslamate.com/tessie
but it fails in reading the zip file from the Tessie export.

Then I tried running the bigquery (link in comment above) on the Tessie export files (after importing them into my Postgres v17 database).
I saved the result in a csv, but importing from that file fails with the error 'Car data incomplete'.

TIA for any advice.
Soeren

**Reply by MattTheTechLV** [reply]
It does appear to be broken to me too. I am also trying to use the "app.myteslamate.com/tessie" converter to convert the data I downloaded from Tessie, so that I can have it in the right format to import into TeslaMate. Unfortunately, no matter what I do, I keep getting the same exact error in the MyTeslaMate Converter. The Error I keep getting is "Error: ZipArchive::close(): Read error: Is a directory"

I am uploading the Zip File that Tessie Gave me. I checked, there is no directory in it, so I am unsure what that error is about. I even tried extracting the data from the Zip, then Re-Zipping it myself with 7zip, and uploading that, same error. I noticed when looking at the Zip that there are a few files with No Data, so I removed those from the Zip and tried again, same error.

I have actually used TeslaMate longer than I have used Tessie, I prefer Teslamate for Data gathering, and I like using Tessie for Controlling the Car from the Browser. But, Unfortunately back in June, I had a Power Outage here in Vegas that lasted 12 hours. I was actually out of town at the time on a business trip. I have Battery Backups on my Server, but the Power Outage unfortunately outlasted the UPS, so I wasn't able to Get the Server back up that runs TeslaMate until I got home from my trip a few days later. 

I say all of this to say that I have about 3 days of Missing Data in TeslaMate. I have just lived with it, but I like Making myself these reports with my charge data, and I finally went to make Last Months Report so I can compare it to this Month's, and realized I am missing several days of data. So, I was hoping to fill in the blanks with the Data From Tessie. I know it polls no where near as often, but I figured something was better than nothing. I downloaded the Data from Tessie for just the 3 days I am missing, but I can't for the life of me get the data converted over to the TeslaFi Format, so I can import it into my TeslaMate.

When I found that MyTeslaMate tool, I thought I could do it without teaching myself how to use Big Query, but I keep running across that error. If I have to, I will learn how to use this Big Query Myself, and then see if I can get the data converted over, but I am not going to have time to teach myself that for at least another week, so I was hoping I could just get the MyTeslaMate converter tool to work, but no luck so far.

Is that the same problem you are having @sbjensen ?

Would love some help on this.

Thank You Everyone!

**Reply by sbjensen** [reply]
Hi

As @MattTheTechLV I’m also missing a few trips and charges due to a loss of power on my server.

I have generated the import file using the BigQuery but it fails during the import in TeslaMate (see my previous post).
I’m using TeslaMate 2.0.

I’m now trying to generate the data for TeslaMate manually but it’s not easy nor advisable if your not into SQL in general and Postgres DB, and don’t have the option of having a test instance of the database.

Regards
Søren

**Reply by jlestel** [reply]
@sbjensen @giterdone509 a field has been renamed by Tessie, it's fixed now: https://github.com/MyTeslaMate/tessie-to-teslamate-import/pull/2

**Reply by MattTheTechLV** [reply]
> @sbjensen @giterdone509 a field has been renamed by Tessie, it's fixed now: [MyTeslaMate/tessie-to-teslamate-import#2](https://github.com/MyTeslaMate/tessie-to-teslamate-import/pull/2)

AWESOME! I just tried the MyTeslaMate/Tessie-toTeslamate-import Website again, and it did give me files to download! So, I don't have to teach myself how to use Big Query Now. 

I have not started the next step of importing into my TeslaMate yet, so I will report back after I try that.

But I am concerned that I will run into the same issues as @sbjensen . He was able to just use Big Query, but then when importing into TeslaMate he got an import fail error. @jlestel will the fix you just implemented also most likely fix the import into TeslaMate?

Regardless I am going to try in a few hours and report back if I was successful. FYI I am also on TeslaMate 2.0, but I am on PostgreSQL Version 16.9, no idea if that makes a difference with this process or not.

Thank You Guys!

**Reply by sbjensen** [reply]
Hi @giterdone509 @MattTheTechLV 

The generation of the TeslaFi file now works again, but TeslaMate does not import the file.
This is a snip of the Teslamate log

teslamate-1  | [info] Found 1 file(s) at '/opt/app/import'. Starting in import mode!
teslamate-1  | 2025-08-01 19:20:42.176 [info] Migrations already up
teslamate-1  | [info] Found 1 file(s) at '/opt/app/import'. Starting in import mode!
teslamate-1  | 2025-08-01 19:20:46.967 [info] No tzdata release files found in custom data dir. Copying release file from tzdata priv dir.
teslamate-1  | 2025-08-01 19:20:47.625 [info] System Info: Erlang/OTP 26 (emu)
teslamate-1  | 2025-08-01 19:20:47.625 [info] Version: 2.0.0
teslamate-1  | PostgreSQL version 17.5 is compatible (17.x series).
teslamate-1  | 2025-08-01 19:20:49.166 [info] POST https://auth.tesla.com/oauth2/v3/token -> 200 (1149.609 ms)
teslamate-1  | 2025-08-01 19:20:49.167 [info] Refreshed api tokens
teslamate-1  | 2025-08-01 19:20:49.185 [info] Scheduling token refresh in 6 h
teslamate-1  | 2025-08-01 19:20:49.214 [info] Running TeslaMateWeb.Endpoint with cowboy 2.13.0 at :::4000 (http)
teslamate-1  | 2025-08-01 19:20:49.222 [info] Access TeslaMateWeb.Endpoint at http://localhost
teslamate-1  | 2025-08-01 19:20:50.841 [info] tzdata release in place is from a file last modified Thu, 05 Sep 2024 18:47:42 GMT. Release file on server was last modified Sat, 22 Mar 2025 21:45:31 GMT.
teslamate-1  | 2025-08-01 19:20:57.739 [info] Tzdata has updated the release from 2024b to 2025b
teslamate-1  | 2025-08-01 19:25:49.486 [info] Update available: 2.0.0 -> 2.1.0
teslamate-1  | 2025-08-01 19:26:07.272 [info] GET /import
teslamate-1  | 2025-08-01 19:26:07.342 [info] Sent 200 in 69ms
teslamate-1  | 2025-08-01 19:26:14.725 [info] Importing 1 file(s) ...
teslamate-1  | 2025-08-01 19:26:14.783 [info] Starting logger for 'Model Y'
teslamate-1  | 2025-08-01 19:26:15.800 [info] Import complete!
teslamate-1  | 2025-08-01 19:26:59.784 [info] GET /import
teslamate-1  | 2025-08-01 19:26:59.809 [info] Sent 200 in 24ms
teslamate-1  | 2025-08-01 19:27:01.788 [info] GET /
teslamate-1  | 2025-08-01 19:27:01.807 [info] Sent 302 in 18ms
teslamate-1  | 2025-08-01 19:27:01.818 [info] GET /import
teslamate-1  | 2025-08-01 19:27:01.844 [info] Sent 200 in 24ms

It says import complete, but no rows shos up in the database/Grafana dashboards.

**Comment by sbjensen** [comment]
Hi

The import page (https://app.myteslamate.com/tessie) seems to be broken again.
Reports the 'is a directory' error again.


Regard Søren

**Reply by muffstic** [reply]
I was able to get it to work with a vehicle that's been archived for a year and a half, but the import fails on current data.

**Reply by flatemwhipit** [reply]
> I was able to get it to work with a vehicle that's been archived for a year and a half, but the import fails on current data.

can confirm this is still happening

**Comment by xdenser** [comment]
Hi, is there any chance to import overlapping data?
My TeslaMate did not log for some time due to recent HTTPS/API/token problems, I was going on road trip just the day it broke so could not update to 4.0.1. Have reenabled tessie as a backup. Now I have a gap of one week (actaully 2 days). Was able to import older data from year 2024 (big 6 kkm roadtrip)  - many thanks for the import function, but this week is missing.

**Comment by ThomasGrl** [comment]
Hi everyone 👋

I ran into the same issues trying to migrate from Tessie to TeslaMate. The web converter at myteslamate.com keeps breaking when Tessie changes its export format, and the BigQuery approach is complex to set up.

So I wrote a **standalone Python script** that converts the Tessie export CSVs directly to TeslaFi format, without any external service or BigQuery dependency:

👉 https://github.com/ThomasGrl/tessie-to-teslamate-converter

**How it works:**
1. Export your data from Tessie (all 5 CSV files)
2. Run the script locally: `python converter.py --input-dir ./tessie --output-dir ./output --vin "YOUR_VIN"`
3. Import the generated `TeslaFi*.csv` files into TeslaMate (select "Etc/UTC" timezone)

**Key features:**
- Pure Python, no external services, no cloud dependencies
- Works offline on your own machine
- Handles AC/DC session classification (peak power > 15 kW = DC)
- Proper cumulative `charge_energy_added` from Energy Remaining
- All the quirks TeslaMate requires (`charger_phases` consistency, required `id` column, etc.)
- You provide your own VIN and display name as arguments — nothing hardcoded

So far it's working well for my Model 3. If you're stuck with the web tools or the BigQuery approach, give it a try — and feel free to open issues or PRs on the repo if something doesn't work with your data!

Also posted this as a feature idea on the TeslaMate repo for native support: https://github.com/teslamate-org/teslamate/discussions/5636

**Comment by JakobLichterfeld** [comment]
#4432
